sunpy/data/data_manager/storage.py

```diff
@@ -0,0 +1,176 @@
+    """
+    Base class for remote data manager storage providers.
+    """
+    @abstractmethod
+    def find_by_key(self, key, value):
+        """
+        Returns the file details if value coresponding to the key
+        found in storage. Returns `None` if hash not found.
+
+        Parameters
+        ----------
+        key: `str`
+            The key/coloumn name of the field.
+        value: `str`
+            The value associated with the key of the entry.
+
+        Returns
+        -------
+        `dict` or `None`
+            `dict` contains the details of the file. `None` if hash not found.
+
+        Raises
+        ------
+        ``KeyError``
+             KeyError is raised if key does not exist.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def delete_by_key(self, key, value):
+        """
+        Deletes the matching entry from the store.
+
+        Parameters
+        ----------
+        key: `str`
+            The key/coloumn name of the field.
+        value: `str`
+            The value associated with the key of the entry.
+
+        Raises
+        ------
+        ``KeyError``
+             KeyError is raised if key does not exist.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def store(self, details):
+        """
+        Stores the details in the storage.
+
+        Parameters
+        ----------
+        details: `dict`
+            Details to be stored.
+        """
+        raise NotImplementedError
+
+
+class InMemStorage(StorageProviderBase):
+    """
+    This provides a storage stored in memory.
+    """
+
+    def __init__(self):
+        self._store = []
+
+    def store(self, details):
+        self._store += [details]
+
+    def delete_by_key(self, key, value):
+        for i in self._store:
+            if i[key] == value:
+                self._store.remove(i)
+
+    def find_by_key(self, key, value):
+        for i in self._store:
+            if i[key] == value:
+                return i
+        return None
+
+
+class SqliteStorage(StorageProviderBase):
+    """
+    This provides a sqlite backend for storage.
+
+    Parameters
+    ----------
+    path: `str`
+        Path to the database file.
+    """
+    COLOUMN_NAMES = [
+        'file_hash',
+        'file_path',
+        'url',
+        'time',
+    ]
+
+    def __init__(self, path):
+        self._db_path = Path(path)
+        self._table_name = 'cache_storage'
+
+        self._db_path.parent.mkdir(parents=True, exist_ok=True)
+        if not self._db_path.exists():
+            # setup database
+            self._setup()
+
+    def _setup(self):
+        schema = ' text, '.join(self.COLOUMN_NAMES) + ' text'
+        try:
+            with self.connection(commit=True) as conn:
+                conn.execute(f'''CREATE TABLE {self._table_name}
+                                ({schema})''')
+        except sqlite3.OperationalError:
+            # Race condition
```

Comment:
race with what?!

Reply:
Tests were causing it. After fixing the tests last week, I'm not sure whether this is needed or not. Maybe I can try removing it and see.